### PR TITLE
Feature/bigger schems

### LINKED ISSUE #1610: [Feature Request] Multi-packet buildings to expand max building size
In discord it was said that the maximum building size is set by the maximum size of a packet you can send through forge. Would it be possible to detect if a scanned building was larger than the maximum packet size, and if so, divide the scan automatically into smaller sections, then piece them together automatically when you build the schematic? Send them one after the other so the preview of the schematic loads in pieces?

Of course, this is a very low priority suggestion AFAIK,

### PR DESCRIPTION
**Pull request by Raycoms**
Closes #1610

# Changes proposed in this pull request:
- Allows to scan bigger schematics
- Limits build tool rendering to 100k blocks.
- Allows to send bigger schematics over to the server and cache them there.

Review please

### CODE REVIEW

File: src/main/java/com/minecolonies/coremod/client/gui/WindowBuildTool.java [hunk 1]
```diff
@@ -824,8 +824,27 @@ private static void requestScannedSchematic(@NotNull final StructureName structu
                 final InputStream stream = Structure.getStream(structureName.toString());
                 if (stream != null)
                 {
-                    Log.getLogger().info("BuilderTool: sending schematic " + structureName + "(md5:" + md5 + ") to the server");
-                    MineColonies.getNetwork().sendToServer(new SchematicSaveMessage(Structure.getStreamAsByteArray(stream)));
+                    //todo!!!!
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/coremod/client/gui/WindowBuildTool.java]: INFO Complete the task associated to this TODO comment. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1135)

File: src/main/java/com/minecolonies/coremod/colony/Structures.java [hunk 2]
```diff
@@ -590,6 +595,94 @@ public static void setMD5s(final Map<String, String> md5s)
...
+     * Handle a schematic which has been cut into pieces.
+     * This method is valid on the server
+     * The schematic will be gathered until all pieces have been put together and then handled like on the client.
+     *
+     * @param bytes representing the schematic.
+     * @param id UUID.
+     * @param piece the piece.
+     * @param pieces the amount of pieces.
+     */
+    public static boolean handleSaveSchematicMessage(final byte[] bytes, final UUID id, final int pieces, final int piece)
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/coremod/colony/Structures.java]: MAJOR Reduce the number of returns of this method 7, down to the maximum allowed 6. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1142)
MAJOR Refactor this method to reduce its Cognitive Complexity from 28 to the 20 allowed. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)

File: src/main/java/com/minecolonies/coremod/colony/Structures.java [hunk 3]
```diff
@@ -282,7 +287,7 @@ private static boolean isSchematicSizeValid(@NotNull final String structureName)
         final byte[] data = Structure.getStreamAsByteArray(Structure.getStream(structureName));
         final byte[] compressed = Structure.compress(data);
 
-        if (compressed != null && compressed.length > maxSize)
+        if (compressed == null)
         {
             Log.getLogger().warn("Structure " + structureName + " is " + compressed.length + " bytes when compress, maximum allowed is " + maxSize + " bytes.");
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/coremod/colony/Structures.java]: MAJOR A "NullPointerException" could be thrown; "compressed" is nullable here. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2259)

File: src/main/java/com/minecolonies/coremod/colony/Structures.java [hunk 4]
```diff
@@ -590,6 +595,94 @@ public static void setMD5s(final Map<String, String> md5s)
...
+                return false;
+            }
+            Log.getLogger().info("Recieved piece: " + piece + " of: " + pieces + " with the size: " + bytes.length + " and ID: " + id.toString());
+            final Map<Integer, byte[]> schemPieces;
+            if(schematicPieces.containsKey(id))
+            {
+                final Tuple<Long, Map<Integer, byte[]>> schemTuple = schematicPieces.remove(id);
+                schemPieces = schemTuple.getSecond();
+
+                if(MathUtils.nanoSecondsToSeconds(System.nanoTime() - schemTuple.getFirst()) > 60)
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/coremod/colony/Structures.java]: MAJOR Assign this magic number 60 to a well-named constant, and use the constant instead. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS109)

File: src/main/java/com/minecolonies/coremod/network/messages/SchematicRequestMessage.java [hunk 5]
```diff
@@ -2,6 +2,7 @@
 
 import com.minecolonies.api.util.Log;
 import com.minecolonies.coremod.MineColonies;
+import com.minecolonies.coremod.util.StructureWrapper;
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/coremod/network/messages/SchematicRequestMessage.java]: MINOR Remove this unused import 'com.minecolonies.coremod.util.StructureWrapper'. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AUselessImportCheck)

File: src/main/java/com/minecolonies/structures/helpers/Structure.java [hunk 6]
```diff
@@ -56,13 +56,19 @@
 import java.util.zip.GZIPInputStream;
 import java.util.zip.GZIPOutputStream;
 
+import static com.minecolonies.api.util.constant.Constants.MAX_SCHEMATIC_SIZE;
 import static com.minecolonies.api.util.constant.Suppression.RESOURCES_SHOULD_BE_CLOSED;
 
 /**
  * Structure class, used to store, create, get structures.
  */
 public class Structure
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/structures/helpers/Structure.java:65]: MAJOR Split this class into smaller and more specialized ones to reduce its dependencies on other classes from 36 to the maximum authorized 30 or less. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1200)

File: src/main/java/com/minecolonies/structures/helpers/Structure.java [hunk 7]
```diff
@@ -566,9 +578,16 @@ public void renderStructure(@NotNull final BlockPos startingPos, @NotNull final
                 }
                 fakeWorld.setTileEntity(aBlockList.pos, tileentity);
             }
+            i = 0;
 
             for (final Template.BlockInfo aBlockList : blockList)
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/structures/helpers/Structure.java:583]: MAJOR Reduce the total number of break and continue statements in this loop to use at most one. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS135)
MINOR Combine this loop with the one that starts on line 563. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3047)

File: src/main/java/com/minecolonies/coremod/network/messages/SaveScanMessage.java [hunk 8]
```diff
@@ -4,16 +4,14 @@
 import com.minecolonies.coremod.util.ClientStructureWrapper;
 import io.netty.buffer.ByteBuf;
 import io.netty.buffer.ByteBufInputStream;
+import io.netty.buffer.ByteBufOutputStream;
 import net.minecraft.nbt.CompressedStreamTools;
-import net.minecraft.nbt.NBTSizeTracker;
 import net.minecraft.nbt.NBTTagCompound;
 import net.minecraft.network.PacketBuffer;
 import net.minecraftforge.fml.common.network.ByteBufUtils;
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/coremod/network/messages/SaveScanMessage.java]: MINOR Remove this unused import 'net.minecraftforge.fml.common.network.ByteBufUtils'. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AUselessImportCheck)

**Comment by BuildTeamLinter** [comment]
SonarQube analysis reported 68 issues
* ![MAJOR][MAJOR] 61 major
* ![MINOR][MINOR] 4 minor
* ![INFO][INFO] 3 info

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [blocks#](https://github.com/Minecolonies/minecolonies/blob/ab78b26d8c40ec02a8f713dddd1dced452b75d26/src/main/java/com/minecolonies/coremod/blocks#): Add a 'package-info.java' file to document the 'blocks' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [gui#](https://github.com/Minecolonies/minecolonies/blob/ab78b26d8c40ec02a8f713dddd1dced452b75d26/src/main/java/com/minecolonies/coremod/client/gui#): Add a 'package-info.java' file to document the 'gui' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [WindowBuildTool.java#L46](https://github.com/Minecolonies/minecolonies/blob/ab78b26d8c40ec02a8f713dddd1dced452b75d26/src/main/java/com/minecolonies/coremod/client/gui/WindowBuildTool.java#L46): class "WindowBuildTool" has 41 methods, which is greater than the 35 authorized. Split it into smaller classes. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1448)
1. ![MAJOR][MAJOR] [WindowBuildTool.java#L811](https://github.com/Minecolonies/minecolonies/blob/ab78b26d8c40ec02a8f713dddd1dced452b75d26/src/main/java/com/minecolonies/coremod/client/gui/WindowBuildTool.java#L811): Refactor this method to reduce its Cognitive Complexity from 23 to the 20 allowed. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)
1. ![MAJOR][MAJOR] [WindowBuildTool.java#L922](https://github.com/Minecolonies/minecolonies/blob/ab78b26d8c40ec02a8f713dddd1dced452b75d26/src/main/java/com/minecolonies/coremod/client/gui/WindowBuildTool.java#L922): Refactor this method to reduce its Cognitive Complexity from 24 to the 20 allowed. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)
1. ![MAJOR][MAJOR] [render#](https://github.com/Minecolonies/minecolonies/blob/ab78b26d8c40ec02a8f713dddd1dced452b75d26/src/main/java/com/minecolonies/coremod/client/render#): Add a 'package-info.java' file to document the 'render' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [colony#](https://github.com/Minecolonies/minecolonies/blob/ab78b26d8c40ec02a8f713dddd1dced452b75d26/src/main/java/com/minecolonies/coremod/colony#): Add a 'package-info.java' file to document the 'colony' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [Structures.java#L174](https://github.com/Minecolonies/minecolonies/blob/ab78b26d8c40ec02a8f713dddd1dced452b75d26/src/main/java/com/minecolonies/coremod/colony/Structures.java#L174): Either log or rethrow this exception. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1166)
1. ![MAJOR][MAJOR] [buildings#](https://github.com/Minecolonies/minecolonies/blob/ab78b26d8c40ec02a8f713dddd1dced452b75d26/src/main/java/com/minecolonies/coremod/colony/buildings#): Add a 'package-info.java' file to document the 'buildings' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [permissions#](https://github.com/Minecolonies/minecolonies/blob/ab78b26d8c40ec02a8f713dddd1dced452b75d26/src/main/java/com/minecolonies/coremod/colony/permissions#): Add a 'package-info.java' file to document the 'permissions' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)

[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: src/main/java/com/minecolonies/coremod/client/gui/WindowBuildTool.java [hunk 9]
```diff
@@ -824,8 +824,35 @@ private static void requestScannedSchematic(@NotNull final StructureName structu
...
+                    else
+                    {
+                        final int pieces = structureAsByteArray.length / MAX_MESSAGE_SIZE;
+
+                        Log.getLogger().info("BuilderTool: sending: " + pieces + " pieces with the schematic " + structureName + "(md5:" + md5 + ") to the server");
+                        for (int i = 1; i <= pieces; i++)
+                        {
+                            final int start = (i - 1) * MAX_MESSAGE_SIZE;
+                            final int size;
+                            if (i == pieces)
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/coremod/client/gui/WindowBuildTool.java:843]: CRITICAL Refactor this code to not nest more than 5 if/for/while/switch/try statements. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS134)

**Comment by BuildTeamLinter** [comment]
SonarQube analysis reported 67 issues
* ![CRITICAL][CRITICAL] 1 critical
* ![MAJOR][MAJOR] 61 major
* ![MINOR][MINOR] 4 minor
* ![INFO][INFO] 1 info

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [blocks#](https://github.com/Minecolonies/minecolonies/blob/ce77d85e3534084ba9d642c1d2dd236952ec187d/src/main/java/com/minecolonies/coremod/blocks#): Add a 'package-info.java' file to document the 'blocks' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [gui#](https://github.com/Minecolonies/minecolonies/blob/ce77d85e3534084ba9d642c1d2dd236952ec187d/src/main/java/com/minecolonies/coremod/client/gui#): Add a 'package-info.java' file to document the 'gui' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [WindowBuildTool.java#L46](https://github.com/Minecolonies/minecolonies/blob/ce77d85e3534084ba9d642c1d2dd236952ec187d/src/main/java/com/minecolonies/coremod/client/gui/WindowBuildTool.java#L46): class "WindowBuildTool" has 41 methods, which is greater than the 35 authorized. Split it into smaller classes. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1448)
1. ![MAJOR][MAJOR] [WindowBuildTool.java#L811](https://github.com/Minecolonies/minecolonies/blob/ce77d85e3534084ba9d642c1d2dd236952ec187d/src/main/java/com/minecolonies/coremod/client/gui/WindowBuildTool.java#L811): Refactor this method to reduce its Cognitive Complexity from 30 to the 20 allowed. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)
1. ![MAJOR][MAJOR] [WindowBuildTool.java#L930](https://github.com/Minecolonies/minecolonies/blob/ce77d85e3534084ba9d642c1d2dd236952ec187d/src/main/java/com/minecolonies/coremod/client/gui/WindowBuildTool.java#L930): Refactor this method to reduce its Cognitive Complexity from 24 to the 20 allowed. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)
1. ![MAJOR][MAJOR] [render#](https://github.com/Minecolonies/minecolonies/blob/ce77d85e3534084ba9d642c1d2dd236952ec187d/src/main/java/com/minecolonies/coremod/client/render#): Add a 'package-info.java' file to document the 'render' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [colony#](https://github.com/Minecolonies/minecolonies/blob/ce77d85e3534084ba9d642c1d2dd236952ec187d/src/main/java/com/minecolonies/coremod/colony#): Add a 'package-info.java' file to document the 'colony' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [Structures.java#L174](https://github.com/Minecolonies/minecolonies/blob/ce77d85e3534084ba9d642c1d2dd236952ec187d/src/main/java/com/minecolonies/coremod/colony/Structures.java#L174): Either log or rethrow this exception. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1166)
1. ![MAJOR][MAJOR] [buildings#](https://github.com/Minecolonies/minecolonies/blob/ce77d85e3534084ba9d642c1d2dd236952ec187d/src/main/java/com/minecolonies/coremod/colony/buildings#): Add a 'package-info.java' file to document the 'buildings' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [permissions#](https://github.com/Minecolonies/minecolonies/blob/ce77d85e3534084ba9d642c1d2dd236952ec187d/src/main/java/com/minecolonies/coremod/colony/permissions#): Add a 'package-info.java' file to document the 'permissions' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by BuildTeamLinter** [comment]
SonarQube analysis reported 65 issues
* ![CRITICAL][CRITICAL] 1 critical
* ![MAJOR][MAJOR] 60 major
* ![MINOR][MINOR] 3 minor
* ![INFO][INFO] 1 info

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [blocks#](https://github.com/Minecolonies/minecolonies/blob/99725b304466d82cafcf1450d561267363741466/src/main/java/com/minecolonies/coremod/blocks#): Add a 'package-info.java' file to document the 'blocks' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [gui#](https://github.com/Minecolonies/minecolonies/blob/99725b304466d82cafcf1450d561267363741466/src/main/java/com/minecolonies/coremod/client/gui#): Add a 'package-info.java' file to document the 'gui' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [WindowBuildTool.java#L46](https://github.com/Minecolonies/minecolonies/blob/99725b304466d82cafcf1450d561267363741466/src/main/java/com/minecolonies/coremod/client/gui/WindowBuildTool.java#L46): class "WindowBuildTool" has 41 methods, which is greater than the 35 authorized. Split it into smaller classes. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1448)
1. ![MAJOR][MAJOR] [WindowBuildTool.java#L811](https://github.com/Minecolonies/minecolonies/blob/99725b304466d82cafcf1450d561267363741466/src/main/java/com/minecolonies/coremod/client/gui/WindowBuildTool.java#L811): Refactor this method to reduce its Cognitive Complexity from 30 to the 20 allowed. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)
1. ![MAJOR][MAJOR] [WindowBuildTool.java#L930](https://github.com/Minecolonies/minecolonies/blob/99725b304466d82cafcf1450d561267363741466/src/main/java/com/minecolonies/coremod/client/gui/WindowBuildTool.java#L930): Refactor this method to reduce its Cognitive Complexity from 24 to the 20 allowed. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)
1. ![MAJOR][MAJOR] [render#](https://github.com/Minecolonies/minecolonies/blob/99725b304466d82cafcf1450d561267363741466/src/main/java/com/minecolonies/coremod/client/render#): Add a 'package-info.java' file to document the 'render' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [colony#](https://github.com/Minecolonies/minecolonies/blob/99725b304466d82cafcf1450d561267363741466/src/main/java/com/minecolonies/coremod/colony#): Add a 'package-info.java' file to document the 'colony' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [Structures.java#L174](https://github.com/Minecolonies/minecolonies/blob/99725b304466d82cafcf1450d561267363741466/src/main/java/com/minecolonies/coremod/colony/Structures.java#L174): Either log or rethrow this exception. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1166)
1. ![MAJOR][MAJOR] [Structures.java#L286](https://github.com/Minecolonies/minecolonies/blob/99725b304466d82cafcf1450d561267363741466/src/main/java/com/minecolonies/coremod/colony/Structures.java#L286): Remove this useless assignment to local variable "maxSize". [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1854)
1. ![MAJOR][MAJOR] [buildings#](https://github.com/Minecolonies/minecolonies/blob/99725b304466d82cafcf1450d561267363741466/src/main/java/com/minecolonies/coremod/colony/buildings#): Add a 'package-info.java' file to document the 'buildings' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Comment by BuildTeamLinter** [comment]
SonarQube analysis reported 70 issues
* ![CRITICAL][CRITICAL] 1 critical
* ![MAJOR][MAJOR] 62 major
* ![MINOR][MINOR] 6 minor
* ![INFO][INFO] 1 info

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [blocks#](https://github.com/Minecolonies/minecolonies/blob/398cea2778305f62194250d4c6f52508f18c48d6/src/main/java/com/minecolonies/coremod/blocks#): Add a 'package-info.java' file to document the 'blocks' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [gui#](https://github.com/Minecolonies/minecolonies/blob/398cea2778305f62194250d4c6f52508f18c48d6/src/main/java/com/minecolonies/coremod/client/gui#): Add a 'package-info.java' file to document the 'gui' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [WindowBuildTool.java#L46](https://github.com/Minecolonies/minecolonies/blob/398cea2778305f62194250d4c6f52508f18c48d6/src/main/java/com/minecolonies/coremod/client/gui/WindowBuildTool.java#L46): class "WindowBuildTool" has 41 methods, which is greater than the 35 authorized. Split it into smaller classes. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1448)
1. ![MAJOR][MAJOR] [WindowBuildTool.java#L811](https://github.com/Minecolonies/minecolonies/blob/398cea2778305f62194250d4c6f52508f18c48d6/src/main/java/com/minecolonies/coremod/client/gui/WindowBuildTool.java#L811): Refactor this method to reduce its Cognitive Complexity from 30 to the 20 allowed. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)
1. ![MAJOR][MAJOR] [WindowBuildTool.java#L930](https://github.com/Minecolonies/minecolonies/blob/398cea2778305f62194250d4c6f52508f18c48d6/src/main/java/com/minecolonies/coremod/client/gui/WindowBuildTool.java#L930): Refactor this method to reduce its Cognitive Complexity from 24 to the 20 allowed. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)
1. ![MAJOR][MAJOR] [render#](https://github.com/Minecolonies/minecolonies/blob/398cea2778305f62194250d4c6f52508f18c48d6/src/main/java/com/minecolonies/coremod/client/render#): Add a 'package-info.java' file to document the 'render' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [colony#](https://github.com/Minecolonies/minecolonies/blob/398cea2778305f62194250d4c6f52508f18c48d6/src/main/java/com/minecolonies/coremod/colony#): Add a 'package-info.java' file to document the 'colony' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [Structures.java#L174](https://github.com/Minecolonies/minecolonies/blob/398cea2778305f62194250d4c6f52508f18c48d6/src/main/java/com/minecolonies/coremod/colony/Structures.java#L174): Either log or rethrow this exception. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1166)
1. ![MAJOR][MAJOR] [Structures.java#L286](https://github.com/Minecolonies/minecolonies/blob/398cea2778305f62194250d4c6f52508f18c48d6/src/main/java/com/minecolonies/coremod/colony/Structures.java#L286): Remove this useless assignment to local variable "maxSize". [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1854)
1. ![MAJOR][MAJOR] [buildings#](https://github.com/Minecolonies/minecolonies/blob/398cea2778305f62194250d4c6f52508f18c48d6/src/main/java/com/minecolonies/coremod/colony/buildings#): Add a 'package-info.java' file to document the 'buildings' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

File: src/main/resources/assets/minecolonies/lang/en_us.lang [hunk 10]
```diff
@@ -667,3 +667,14 @@ tile.minecolonies.blockTimberFrame_dark_oak_horizontalnocap.name= Dark Oak Horiz
 
 com.minecolonies.coremod.request.toolow=Can't mine block %s at %s. Need to upgrade the building to use a tool with hihgher mining level.
 com.minecolonies.coremod.dimension.no=You are not allowed to place colonies in other dimensions (Triggered by config value)
+
+tile.minecolonies.blockshingleslab_sandstone.name=Sandstone Shingle Slab
```
Comment (marchermans) [inline src/main/resources/assets/minecolonies/lang/en_us.lang]: I think that this does not belong here.

**Review by marchermans — CHANGES_REQUESTED** [review]

**Comment by BuildTeamLinter** [comment]
SonarQube analysis reported 70 issues
* ![CRITICAL][CRITICAL] 1 critical
* ![MAJOR][MAJOR] 62 major
* ![MINOR][MINOR] 6 minor
* ![INFO][INFO] 1 info

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MAJOR][MAJOR] [blocks#](https://github.com/Minecolonies/minecolonies/blob/89865655daa9e6f70ef8c4eda921c3a278e5025b/src/main/java/com/minecolonies/coremod/blocks#): Add a 'package-info.java' file to document the 'blocks' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [gui#](https://github.com/Minecolonies/minecolonies/blob/89865655daa9e6f70ef8c4eda921c3a278e5025b/src/main/java/com/minecolonies/coremod/client/gui#): Add a 'package-info.java' file to document the 'gui' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [WindowBuildTool.java#L46](https://github.com/Minecolonies/minecolonies/blob/89865655daa9e6f70ef8c4eda921c3a278e5025b/src/main/java/com/minecolonies/coremod/client/gui/WindowBuildTool.java#L46): class "WindowBuildTool" has 41 methods, which is greater than the 35 authorized. Split it into smaller classes. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1448)
1. ![MAJOR][MAJOR] [WindowBuildTool.java#L811](https://github.com/Minecolonies/minecolonies/blob/89865655daa9e6f70ef8c4eda921c3a278e5025b/src/main/java/com/minecolonies/coremod/client/gui/WindowBuildTool.java#L811): Refactor this method to reduce its Cognitive Complexity from 30 to the 20 allowed. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)
1. ![MAJOR][MAJOR] [WindowBuildTool.java#L930](https://github.com/Minecolonies/minecolonies/blob/89865655daa9e6f70ef8c4eda921c3a278e5025b/src/main/java/com/minecolonies/coremod/client/gui/WindowBuildTool.java#L930): Refactor this method to reduce its Cognitive Complexity from 24 to the 20 allowed. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)
1. ![MAJOR][MAJOR] [render#](https://github.com/Minecolonies/minecolonies/blob/89865655daa9e6f70ef8c4eda921c3a278e5025b/src/main/java/com/minecolonies/coremod/client/render#): Add a 'package-info.java' file to document the 'render' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [colony#](https://github.com/Minecolonies/minecolonies/blob/89865655daa9e6f70ef8c4eda921c3a278e5025b/src/main/java/com/minecolonies/coremod/colony#): Add a 'package-info.java' file to document the 'colony' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [Structures.java#L174](https://github.com/Minecolonies/minecolonies/blob/89865655daa9e6f70ef8c4eda921c3a278e5025b/src/main/java/com/minecolonies/coremod/colony/Structures.java#L174): Either log or rethrow this exception. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1166)
1. ![MAJOR][MAJOR] [Structures.java#L286](https://github.com/Minecolonies/minecolonies/blob/89865655daa9e6f70ef8c4eda921c3a278e5025b/src/main/java/com/minecolonies/coremod/colony/Structures.java#L286): Remove this useless assignment to local variable "maxSize". [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1854)
1. ![MAJOR][MAJOR] [buildings#](https://github.com/Minecolonies/minecolonies/blob/89865655daa9e6f70ef8c4eda921c3a278e5025b/src/main/java/com/minecolonies/coremod/colony/buildings#): Add a 'package-info.java' file to document the 'buildings' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Pull request merged**